core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -204,7 +211,154 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
         return topicPartitionData;
     }
 
-    private void releasePartitionLocks(String groupId, Set<TopicIdPartition> topicIdPartitions) {
+    // In case, fetch offset metadata doesn't exist for any topic partition in the list of topic partitions, we do a
+    // replicaManager.readFromLog to populate the offset metadata.
+    private MaybeUpdateFetchOffsetMetadataResult maybeUpdateFetchOffsetMetadataForTopicPartitions(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData) {
+        boolean isFetchOffsetMetadataUpdated = false;
+        Map<TopicIdPartition, FetchRequest.PartitionData> missingFetchOffsetMetadataTopicPartitions = new LinkedHashMap<>();
+        Map<TopicIdPartition, FetchPartitionOffsetData> replicaManagerReadResponseData;
+        for (Map.Entry<TopicIdPartition, FetchRequest.PartitionData> entry : topicPartitionData.entrySet()) {
+            TopicIdPartition topicIdPartition = entry.getKey();
+            SharePartition sharePartition = sharePartitionManager.sharePartition(shareFetchData.groupId(), topicIdPartition);
+            if (sharePartition == null) {
+                log.debug("Encountered null share partition for groupId={}, topicIdPartition={}. Skipping it.", shareFetchData.groupId(), topicIdPartition);
+                continue;
+            }
+            if (sharePartition.latestFetchOffsetMetadata().isEmpty())
+                missingFetchOffsetMetadataTopicPartitions.put(topicIdPartition, entry.getValue());
+        }
+
+        if (missingFetchOffsetMetadataTopicPartitions.isEmpty()) {
+            return new MaybeUpdateFetchOffsetMetadataResult(false, null);
+        }
```

Comment:
So this iteration will always be excuted for every share fetch when the `missingFetchOffsetMetadataTopicPartitions` will rarely be true, only when a new SharePartition is created. Hence I was thinking why not to have such update only on SharePartition initilization. Though I understand that current `readFromLog` API requires fethchParams but is there an API which can supply the LogOffsetMetadata when requested with topic partition and specific offset(start offset of share partition)? @junrao wdyt?

Reply:
So this suggestion is not apt as we need the refreshed information https://github.com/apache/kafka/pull/17539#discussion_r1816800887